src/devices/Ak8963/Ak8963.cs

```diff
@@ -0,0 +1,275 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.Device.I2c;
+using System.IO;
+using System.Numerics;
+using System.Threading;
+
+namespace Iot.Device.Ak8963
+{
+    /// <summary>
+    /// AK8963 class implementing a magnetometer
+    /// </summary>
+    public class Ak8963 : IDisposable
+    {
+        private I2cDevice _i2cDevice;
+        private bool _autoDispose;
+        private MeasurementMode _measurementMode;
+        private OutputBitMode _outputBitMode;
+        private bool _selfTest = false;
+        private Vector3 _magnetometerBias = Vector3.Zero;
+        private Ak8963Interface _ak8963Interface;
+
+        public const byte DefaultI2cAddress = 0x0C;
+
+        public Ak8963(I2cDevice i2CDevice, bool autoDispose = true) : this(i2CDevice, new Ak8963I2c(), autoDispose)
+        { }
+
+        public Ak8963(I2cDevice i2CDevice, Ak8963Interface ak8963Interface, bool autoDispose = true)
+        {
+            _i2cDevice = i2CDevice;
+            _autoDispose = autoDispose;
+            _ak8963Interface = ak8963Interface;
```

Comment:
I find this very confusing, I don't think it's right to have to pass this which only defines how to read using i2c. This shouldn't be needed, and the read methods should be implemented in this class.

Reply:
Unfortunately, we have to. The device can be used on the back of another device, in this PR on the back of the MPU9250 and you can't really access it directly. It would have been much easier. So you have to rewrit primitives of I2C access in order to have it working. See src/devices/Mpu9250/Ak8963Attached.cs.

Reply:
So does that mean that the Mpu9250 contains a Ak8963 within itself? When connecting Mpu9250, if you run i2cdump, do you see the Ak8963? If so I still quite don't understand why would you need this.

Reply:
Yes, MPU9250 contains an AK8963. It's just that the MPU9250 is a master I2C with slave devices. So to access those slave devices, you need to go thru the MPU9250. In theory, there is a mode to make the slave devices visible to the outside world, practically, I never managed to make it working correctly, not even having the AK8963 visible. So the communication to slave devices go thru the MPU9250 thru I2C before going to the slave devices thru another I2C bus from  the MPU9250

Reply:
Also we talked offline with @Ellerbach about exposing I2C of MPU9250 but seems that "virtual" I2C exposed MPU9250 is not compatible with I2cDevice. Seems like MPU9250's I2C is register oriented rather than bytes oriented. It feels like we are lacking this abstraction somewhere. Long term we should design such abstraction but we've decided to better keep it off this PR for simplicity